cpp/src/arrow/error_or.h

```diff
@@ -0,0 +1,457 @@
+                                                 std::is_convertible<U, T>::value>::type>
+  ErrorOr(const ErrorOr<U>& other) : has_value_(other.has_value_) {
+    if (has_value_) {
+      // Re-use the memory from variant when constructing
+      new (&variant_) variant(other.variant_.value_);
+    } else {
+      new (&variant_) variant(other.variant_.status_);
+    }
+  }
+
+  /// Copy-assignment operator.
+  ///
+  /// \param other The ErrorOr object to copy.
+  ErrorOr& operator=(const ErrorOr& other) {
+    // Check for self-assignment.
+    if (this == &other) {
+      return *this;
+    }
+
+    // Construct the variant object using the variant object of the source.
+    if (other.has_value_) {
+      AssignValue(other.variant_.value_);
+    } else {
+      AssignStatus(other.variant_.status_);
+    }
+    return *this;
+  }
+
+  /// Templatized constructor which constructs a `ErrorOr<T>` by moving the
+  /// contents of a `ErrorOr<U>`. `T` must be implicitly constructible from `U
+  /// &&`.
+  ///
+  /// Sets `other` to contain a non-OK status with a`StatusError::Invalid`
+  /// error code.
+  ///
+  /// \param other The ErrorOr object to move from and set to a non-OK status.
+  template <typename U,
+            typename E = typename std::enable_if<std::is_constructible<T, U>::value &&
+                                                 std::is_convertible<U, T>::value>::type>
+  ErrorOr(ErrorOr<U>&& other) : has_value_(other.has_value_) {
+    if (has_value_) {
+      // Re-use the memory from variant when constructing
+      new (&variant_) variant(std::move(other.variant_.value_));
+#ifndef NDEBUG
+      other.OverwriteValueWithStatus(
+          Status::Invalid("Value was moved to another ErrorOr."));
+#else
+      new (&other.variant_) variant(Status::OK());
+      other.has_value_ = false;
+#endif
+    } else {
+      // Re-use the memory from variant when constructing
+      new (&variant_) variant(std::move(other.variant_.status_));
+#ifndef NDEBUG
+      // The other.variant_.status_ gets moved and invalidated with a Status-
+      // specific error message above. To aid debugging, set the status to a
+      // ErrorOr-specific error message.
+      other.variant_.status_ = Status::Invalid("status was moved to another ErrorOr.");
+#endif
+    }
+  }
+
+  /// Move-assignment operator.
+  ///
+  /// Sets `other` to contain a non-OK status.
+  ///
+  /// \param other The ErrorOr object to assign from and set to a non-OK
+  /// status.
+  ErrorOr& operator=(ErrorOr&& other) {
+    // Check for self-assignment.
+    if (this == &other) {
+      return *this;
+    }
+
+    // Construct the variant object using the variant object of the donor.
+    if (other.has_value_) {
+      AssignValue(std::move(other.variant_.value_));
+#ifndef NDEBUG
+      other.OverwriteValueWithStatus(
+          Status::Invalid("Value was moved to another ErrorOr."));
+#else
+      new (&other.variant_) variant(Status::OK());
```

Comment:
Need to double check if a call destructor here is necessary and in other #elsr blocks

Reply:
In general: since we already have an independently tested (vendored) variant class, is it worthwhile to just refactor to use that?

Reply:
Originally I was against that, but I agree and this is now done.